src/bit_copy.cc

```diff
@@ -0,0 +1,230 @@
+// Copyright 2018 The Amber Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/bit_copy.h"
+
+#include <cassert>
+
+namespace amber {
+
+// static
+Result BitCopy::CopyValueToBuffer(uint8_t* dst,
+                                  const Value& src,
+                                  uint8_t dst_bit_offset,
+                                  uint8_t bits) {
+  uint64_t data = 0;
+  if (src.IsInteger()) {
+    if (bits <= 8) {
+      uint8_t* ptr = reinterpret_cast<uint8_t*>(&data);
+      *ptr = src.AsUint8();
+    } else if (bits <= 16) {
+      uint16_t* ptr = reinterpret_cast<uint16_t*>(&data);
+      *ptr = src.AsUint16();
+    } else if (bits <= 32) {
+      uint32_t* ptr = reinterpret_cast<uint32_t*>(&data);
+      *ptr = src.AsUint32();
+    } else if (bits <= 64) {
+      data = src.AsUint64();
+    } else {
+      return Result("Invalid int bits for CopyValueToBuffer");
+    }
+  } else {
+    if (bits == 64) {
+      double* ptr = reinterpret_cast<double*>(&data);
+      *ptr = src.AsDouble();
+    } else {
+      float* float_ptr = nullptr;
+      uint16_t* uint16_ptr = nullptr;
+      switch (bits) {
+        case 32:
+          float_ptr = reinterpret_cast<float*>(&data);
+          *float_ptr = src.AsFloat();
+          break;
+        case 16:
+        case 11:
+        case 10:
+          uint16_ptr = reinterpret_cast<uint16_t*>(&data);
+          *uint16_ptr = FloatToHexFloat(src.AsFloat(), bits);
+          break;
+        default:
+          return Result("Invalid float bits for CopyValueToBuffer");
+      }
+    }
+  }
+
+  while (dst_bit_offset > 7) {
+    ++dst;
+    dst_bit_offset -= 8;
+  }
+
+  // No overflow will happen. |dst_bit_offset| is based on VkFormat
+  // and if |bits| is 64, |dst_bit_offset| must be 0. No component
+  // has |bits| bigger than 64.
+  data <<= dst_bit_offset;
+
+  CopyBits(dst, data, dst_bit_offset, bits);
+  return {};
+}
+
+// static
+void BitCopy::CopyMemoryToBuffer(uint8_t* dst,
```

Comment:
This is only used by verifyer, why not put it in the anonymous namespace of verifier.cc?